interfaces/utils/path_patterns_test.go

```diff
@@ -0,0 +1,130 @@
+
+func (s *pathPatternsSuite) TestRegexCreationHappy(c *C) {
+	// to save some typing:
+	d := utils.GlobDefault
+	n := utils.GlobNull
+
+	data := []struct {
+		pattern       string
+		glob          utils.GlobFlags
+		expectedRegex string
+	}{
+		{`/media/user/`, d, `^/media/user/$`},
+		{`/dev/sd*`, d, `^/dev/sd[^/\x00]*$`},
+		{`/dev/sd*`, n, `^/dev/sd[^/]*$`},
+		{`/dev/sd?`, d, `^/dev/sd[^/\x00]$`},
+		{`/etc/**`, d, `^/etc/[^/\x00][^\x00]*$`},
+		{`/home/*/.bashrc`, d, `^/home/[^/\x00][^/\x00]*/\.bashrc$`},
+		{`/home/*/.bashrc`, n, `^/home/[^/][^/]*/\.bashrc$`},
+		{`/media/{user,loser}/`, d, `^/media/(user|loser)/$`},
+		{`/nested/{a,b{c,d}}/`, d, `^/nested/(a|b(c|d))/$`},
+		{`/media/\{in-braces\}/`, d, `^/media/\{in-braces\}/$`},
+		{`/media/\[in-brackets\]/`, d, `^/media/\[in-brackets\]/$`},
+		{`/dev/sd[abc][0-9]`, d, `^/dev/sd[abc][0-9]$`},
+		{`/quoted/bracket/[ab\]c]`, d, `^/quoted/bracket/[ab\]c]$`},
+		{`{[,],}`, d, `^([,]|)$`},
+		{`/path/with/comma[,]`, d, `^/path/with/comma[,]$`},
+	}
+
+	for _, testData := range data {
+		pattern := testData.pattern
+		expectedRegex := testData.expectedRegex
+		regex, err := utils.CreateRegex(pattern, testData.glob)
+		c.Assert(err, IsNil, Commentf("%s", pattern))
+		c.Assert(regex, Equals, expectedRegex, Commentf("%s", pattern))
+		// Also, make sure that the obtained regex is valid
+		_, err = regexp.Compile(regex)
+		c.Assert(err, IsNil, Commentf("%s", pattern))
+	}
+}
+
+func (s *pathPatternsSuite) TestRegexCreationUnhappy(c *C) {
+	data := []struct {
+		pattern       string
+		expectedError string
+	}{
+		{`/media/{}/`, `Invalid number of items between {}:.*`},
+		{`/media/{some/things`, `Missing 1 closing brace\(s\):.*`},
+		{`/media/}`, `Invalid closing brace, no matching open { found:.*`},
+		{`/media/[abc`, `Missing closing bracket ']':.*`},
+		{`/media/]`, `Pattern contains unmatching ']':.*`},
+		{`/media\`, `Expected character after '\\':.*`},
+		// 123456789012345678901234567890123456789012345678901, 51 of them
+		{`/{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{`, `Maximum group depth exceeded:.*`},
+		{`/comma/not/in/group/a,b`, `Cannot use ',' outside of group or character class`},
+	}
+
+	for _, testData := range data {
+		pattern := testData.pattern
+		expectedError := testData.expectedError
+		pathPattern, err := utils.NewPathPattern(pattern)
+		c.Assert(pathPattern, IsNil, Commentf("%s", pattern))
+		c.Assert(err, ErrorMatches, expectedError, Commentf("%s", pattern))
+	}
+}
+
+func (s *pathPatternsSuite) TestPatternMatches(c *C) {
+	data := []struct {
+		pattern       string
+		testPath      string
+		expectedMatch bool
+	}{
+		{`/same/path/`, `/same/path/`, true},
+		{`/path/*`, `/path/here`, true},
+		{`/path/*`, `/path/too/deep`, false},
+		{`/path/**`, `/path/here`, true},
+		{`/path/**`, `/path/here/too`, true},
+		{`/dev/sd?`, `/dev/sda`, true},
+		{`/dev/sd?`, `/dev/sdb1`, false},
+		{`/media/{user,loser}/`, `/media/user/`, true},
```

Comment:
whatever the result of the `^` and `$` question I asked above is, we should have those covered in tests here either in the happy paths or in the unhappy paths

Reply:
Added